packages/storage/src/chrome/base.ts

```diff
@@ -23,26 +18,36 @@ type Version = string;
 // See `migration.test.ts` for an example.
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
 
-export class ExtensionStorage<T> {
+export class ExtensionStorage<T, D extends keyof T> {
   constructor(
-    private storage: IStorage,
-    private defaults: T,
+    private storage: chrome.storage.StorageArea,
+    private defaults: Pick<T, D>,
     private version: Version,
-    private migrations: Migrations<keyof T>,
+    private migrations: Partial<Record<keyof T, Migration>>,
   ) {}
 
-  async get<K extends keyof T>(key: K): Promise<T[K]> {
+  async get<K extends keyof T>(key: K): Promise<Partial<T>[K]> {
+    console.log('GET', key);
     const result = (await this.storage.get(String(key))) as
       | Record<K, StorageItem<T[K]>>
       | EmptyObject;
 
-    if (isEmptyObj(result)) return this.defaults[key];
-    else return await this.migrateIfNeeded(key, result[key]);
+    // no value in storage
+    if (isEmptyObj(result)) {
+      if (key in this.defaults) return this.defaults[key as K & D];
+      else return (result as Partial<T>)[key as Exclude<K, D>];
+    }
+
+    // old version in storage
+    if (result[key].version !== this.version) return await this.migrate(key, result[key]);
+
+    // normal case
+    return result[key].value;
   }
```

Comment:
ideally, any PR doing this would type the getter such that keys in `this.defaults` are guaranteed and don't need a nullish check by the caller